src/components/bulk-email-tool/bulk-email-form/BulkEmailForm.jsx

```diff
@@ -267,6 +269,44 @@ function BulkEmailForm(props) {
         }}
       />
       <Form>
+        {/* this will be pluggable */}
+        <div className="border border-brand-400 p-3">
+          <h1>Pluggable components</h1>
+          <PluggableComponent
+            id="checkbox-form"
+            as="communications-app-check-box-form"
+            label="checkbox label - @openedx-plugins/communications-app-check-box-form"
+            isChecked
```

Comment:
It seems a bit strange to me that these plugins are so specific. In this case a checkbox and an input box. What if I want to replace the check box with two checkboxes, or with an input box? It seems like that sort of flexibility is missing here. With props like `isChecked` being passed to the plugin, it seems like the checkbox plugin must render a checkbox. So isn't the customization too limited?

Also, I guess you're still working on this but it seems like `handleCheck` is missing.

Reply:
I agree. If you have any number of children, the props will be limited only to the plugin. However, we would need to pass props for each child in a prop called 'childrenProps' or something similar, which would be an object to specify its props for each one. But I don't like that the component's props will be too many if you have many components as children. I think each plugin should be responsible for a specific task. I was wondering if there is another way to handle that customization case

Reply:
I think in general the plugins should get a very simple prop that applies no matter how many widgets are in the slot, or they should get their data from React Context, not props.

Here is one idea for how you could implement something like that:

```ts
function BulkEmailForm(props) {

// Store the state of all the form fields. The fields are controlled by plugins.
// Each key in formFieldsState is a plugin ID, and each value is {value: any, isReady: boolean}
const [formFieldsState, updateFormFieldsState] = React.useState({});
const isReady = Object.keys(formFieldsState).length > 1 && Object.values(formFieldsState).every(field => field.isReady);
const pluginProps = {formFieldsState, updateFormFieldsState};

return <Form>
    <PluggableComponents id="bulk-email-form" {...pluginProps}>
        <PluggableComponent id="bulk-email-form-email-recipient" {...pluginProps}>
        <PluggableComponent id="bulk-email-form-email-subject" {...pluginProps}>
        <PluggableComponent id="bulk-email-form-email-body" {...pluginProps}>
    </PluggableComponents>
</Form>;
}
```

Then each plugin can manage its own state (checked or not) and data like this:
```
const pluginId = "new-checkbox-field";
const state = props.formFieldsState[pluginid] ?? {};
const updateState = (changes) => {
    props.updateFormFieldsState(oldState => ({...oldState, [pluginId]: {...oldState[pluginId], ...changes}));
};
if (!state.isReady) {
    updateState({isReady: true}); // The checkbox is not required, so allow users to submit the form whether or not this is checked.
}

return <Checkbox checked={state.value} onChange={(isChecked) => updateState({value: isChecked})} />
```

Reply:
Awesome! This is a really great idea, I will check this out. That way even each pluggableComponents can update the state of the parent and sometimes you will need to access to that data. Thanks a lot I'll be refactoring all the form that way

Reply:
Was this addressed, yet?  Just checking.